common/nodes/database-schema.xml

```
@@ -466,6 +466,32 @@ CREATE TABLE firmware_mapping (
 	CONSTRAINT unique_node_firmware UNIQUE (node_id, firmware_id)
 );
 
+<!-- Virtual Machines -->
+CREATE TABLE virtual_machines (
+	id INT AUTO_INCREMENT PRIMARY KEY,
+	hypervisor_id INT DEFAULT NULL,
+	node_id INT NOT NULL,
+	memory_size INT NOT NULL,
+	cpu_cores INT NOT NULL,
+	vm_delete BOOLEAN DEFAULT 0,
+	FOREIGN KEY (hypervisor_id) REFERENCES nodes(id) ON DELETE SET NULL,
```

Does this make sense to set null if the hypervisor node is deleted? I would think this should instead have no `ON DELETE` clause so if someone tries to delete the hypervisor before deleting the VMs on it, the delete would fail. This is based on the assumption that a virtual_machine has resources associated with it outside of the DB that would need to be cleaned up before the DB entry was removed.

---

Thinking about it, I think you are right, when you remove a hypervisor with any virtual machines assigned to it, remove host should raise an error along with some schema level enforcement as well. This would require the user to either reassign the vm to another hypervisor (if this should be allowed without cleanup as well I'll discuss more in your other comment) or remove the virtual machines and sync them first.